src/utils/item.js

```diff
@@ -0,0 +1,93 @@
+import { getDaysBetweenDates } from './dates';
+
+/*sorting logic for items*/
+
+let currentTime = new Date().getTime(); //in MS
+
+export const checkIfActive = (ref, current) => {
+	return getDaysBetweenDates(ref, current) < 60;
+};
+
+export const sortByNextDateAlphabetical = (array) => {
+	return array.sort((a, b) => {
+		return a.daysToNext - b.daysToNext || a.name.localeCompare(b.name);
+	});
+};
+
+export function comparePurchaseUrgency(data) {
+	/**with the following behaviors
+	sorts inactive items last, then
+	sorts items in ascending order of days until purchase, and
+	sorts items with the same days until purchase alphabetically */
+
+	const dataWithDays = data.map((item) => {
+		// we will use these frequently in the next ops, so passing as additional keys:
+		let refTime = item.dateLastPurchased
+			? item.dateLastPurchased.toMillis()
+			: item.dateCreated.toMillis();
+
+		let daysToNext = getDaysBetweenDates(
+			currentTime,
+			item.dateNextPurchased.toMillis(),
+		);
+		return { ...item, refTime, daysToNext };
+	});
+
+	const sortedOverdueList = sortByNextDateAlphabetical(
+		dataWithDays
+			.filter((item) => {
+				const { refTime, daysToNext } = item;
+				return checkIfActive(refTime, currentTime) && daysToNext < 0;
+			})
+			.map((item) => {
+				let urgency = 'overdue';
+				let urgencyMessage = 'Overdue';
+				return { ...item, urgency, urgencyMessage };
+			}),
+	);
+
+	const sortedActiveListNoOverdues = sortByNextDateAlphabetical(
+		dataWithDays
+			.filter((item) => {
+				const { refTime, daysToNext } = item;
+				return checkIfActive(refTime, currentTime) && daysToNext >= 0;
+			})
+			.map((item) => {
+				let urgency;
+				let urgencyMessage;
+
+				const { daysToNext } = item;
+				if (daysToNext < 0) {
+					urgency = 'overdue';
+					urgencyMessage = 'Overdue';
+				} else if (daysToNext <= 7) {
+					urgency = 'soon';
+					urgencyMessage = 'Soon';
+				} else if (daysToNext > 7 && daysToNext < 30) {
+					urgency = 'kind-of-soon';
+					urgencyMessage = 'Kind Of Soon';
+				} else if (daysToNext >= 30) {
+					urgency = 'not-soon';
+					urgencyMessage = 'Not Soon';
+				}
+
+				return { ...item, urgency, urgencyMessage };
+			}),
+	);
+	const sortedInactiveListByDateNext = sortByNextDateAlphabetical(
+		dataWithDays
+			.filter((item) => {
+				const { refTime } = item;
+				return !checkIfActive(refTime, currentTime);
+			})
+			.map((item) => {
+				let urgency = 'inactive';
+				let urgencyMessage = 'Inactive';
+				return { ...item, urgency, urgencyMessage };
+			}),
+	);
+	return sortedOverdueList.concat(
+		sortedActiveListNoOverdues,
+		sortedInactiveListByDateNext,
+	);
+}
```

Comment:
Couple ideas and questions here:
- You probably want to break this apart into separate functions here for both readability and testability. If you want them to be "private" to this function, you could omit the `export` for clarity on that.
- Separating the "sorting" from the "adding urgency labels" makes it so you don't rewrite the same urgency label code.
- I think it might be unnecessary to add 2 separate "urgency" keys on the data rather than just one. I know you're using it to reference a CSS class in the `ListItem` component, but maybe there's a different way. What I would suggest is to change the `urgency` to a `number` value instead, and keeping the Urgency Message data in `ListItem`. So you could set something like `urgency = 0` and key that to an `urgencyMessage[0]` in  a list of messages there.

Here's a sorting function I set up that uses a multi-array, which is similar to what you're doing here with `concat`.
```suggestion
const appendDataToDays = (data) =>
	data.map((item) => {
		let refTime = item.dateLastPurchased
			? item.dateLastPurchased.toMillis()
			: item.dateCreated.toMillis();

		let daysToNext = getDaysBetweenDates(
			currentTime,
			item.dateNextPurchased.toMillis(),
		);
		return { ...item, refTime, daysToNext };
	});

const sortedMultiArr = (dataWithDays) =>
	dataWithDays.reduce(
		(sorted, item) => {
			if (!checkIfActive(item.refTime, currentTime)) {
				sorted[2].push(item);
			} else if (
				checkIfActive(item.refTime, currentTime) &&
				item.daysToNext < 0
			) {
				sorted[0].push(item);
			} else sorted[1].push(item);
			return sorted;
		},
		[[], [], []],
	);

const addUrgency = (item) => {
	let urgency;
	let urgencyMessage;

	const { daysToNext } = item;
	if (daysToNext < 0) {
		urgency = 'overdue';
		urgencyMessage = 'Overdue';
	} else if (daysToNext <= 7) {
		urgency = 'soon';
		urgencyMessage = 'Soon';
	} else if (daysToNext > 7 && daysToNext < 30) {
		urgency = 'kind-of-soon';
		urgencyMessage = 'Kind Of Soon';
	} else if (daysToNext >= 30) {
		urgency = 'not-soon';
		urgencyMessage = 'Not Soon';
	}

	return { ...item, urgency, urgencyMessage };
};

export const comparePurchaseUrgency = (data) => {
	const dataWithDays = appendDataToDays(data);
	const sorted = sortedMultiArr(dataWithDays)
		.map(sortByNextDateAlphabetical)
		.flat();

	return sorted.map(addUrgency);
}

```

So in this example, the main function operates like this:
1.  Append data with ref keys
2. Pass that data to a function that splits overdue, active, inactive lists into a multi array in correct order
3. Each sub array in the multi array gets sorted with `sortByNextDateAlphabetical`
4. Multiarray gets flattened
5. Add urgency keys to everything
6. Return data

Reply:
Thanks for the suggestions! I see what you mean. For rationale,
- I intentionally made the "sub" functions exportable for testing, so that was why I broke them out! Not familiar enough with jest but maybe there is a better way.
- Thought of breaking out the key-attaching step as a function but did not find a compelling reason to do that rather than just mapping directly. I didn't have a strong POV there though!
- Would not have thought of combining .reduce() with a multi-array approach! Definitely more succinct - and I do like that it goes through each item just once for that step.
- I think we were using the `urgency` variable to double as a class name, but can also see that becoming more succinct too with multi-arrays where each sub-array consists of both intended class name and urgency message. And maybe close to your initial idea @andentx ?

Reply:
Oh what I meant was that you could break down the sub functions in `comparePurchaseUrgency` further, and was thinking that the reason you might have not was because you might have wanted only that function to have access to it. For "sub functions" I mean the current smaller functions happening inside your `comparePurchaseUrgency` that weren't exported:

 1 -  add extra keys to data
 2 -  filter data into separate groups by activity/inactivity
 4 - add urgency to all keys
 
I'll also note that a lot of times you might want to break functions out of the parent function for readability, even if the smaller function isn't going to be used anywhere else. That way the parent function acts more like a roadmap,leading 
 you to the details abstracted into children functions if you're looking for that.

Reply:
hey!
1 - yes, that's the one that I mentioned is called just once - so I was ambivalent about the value of breaking that out either way
2 - filtering data into large groups: I liked your solution and thought the reduce() x 1 runthrough was definitely more efficient
3 - Ah, I was agreeing re: maybe passing a single urgency number to serve as an index, and do any labeling (including potential array) in the ListItem. But also works re: just using daysToNext to do that in ListItem. This also makes keeping those extra keys feel less redundant.

Reply:
Still WIP as the suggestion doesn't include 'Inactive' labeling - currently approaching labeling inside ListItem.

Reply:
Ooops sory I must have missed that one! 
In which case does it get marked 'Inactive' by the way? Was curious as I couldn't tell off hand what that case was describing.

Reply:
Oops sorry must have missed that one! You could probably just add another check in the `addUrgency` fn for that one, though I know you've made a `checkIfActive` function already.

Reply:
Yeah, have been tinkering but for some reason the conditionals are being a bit tricky. Will push updates when it starts working.

Reply:
Updated and shifted urgency labeling into ListItem!

Reply:
> re: `urgency` label and nested arrays 
...
> That way everything to do with finite greetings or data is easily accessible in the parent component it's related to.


Great suggestions! 

Using both `urgency` and `urgencyMessage` to represent essentially the same info/data definitely felt wrong. I was hoping to clean it up later but wasn’t sure how until now!  